### ISSUE TITLE
Identity recover

### ISSUE THREAD

**Issue by strk** [friendica/friendica#3466]
I've [learn](https://github.com/friendica/friendica/pull/3439#issuecomment-301532764) that loosing the private key of your identity makes it impossible to recover. If this is confirmed we need a way to allow recovery.

**Comment by fabrixxm** [comment]
- Go to /uexport (Settings -> Export personal data).
- Select "Export account".
- Save the file somewere.
*time passes, your server explode, you reinstall it but you don't have backups*
- Go to /uimport (Register -> Import account)

**Comment by strk** [comment]
That's equivalent of having a backup.
This ticket is to provide a way to allow your friends to still
reference you by your previous handle, just confirming your updated
keypair.

But surely having a YELLOW WARNING STRIPE until you confirm you
took that export would be a step foward.

**Comment by Hypolite** [comment]
If you don't have the private key then there's no way of recovering your account someone else couldn't use to impersonate you. That's the whole point of public-private key pairs.

@fabrixxm is right about the only way of preventing such issue in the first place: backing up your private key so that you always have it.

So we can't add a way of recovering your account (because there's no trusted third party), however we can make it obvious or mandatory to backup the user account once created.

**Comment by strk** [comment]
@hypolite I hardly disagree about leaving the choice to the software and not to the user. You have to decide if you want to trust the new user key or not. It is up to you to find an iffline channel if needed.

Do you trust my profile URL to be under my control ?

We used to be "friends", don't you find it absurd that your software gets to decide we cannot be anymore now ?

**Comment by annando** [comment]
If you find an offline channel where you can verify the other person then both people can decide to remove the contact on their side and to join again.

Of course this is much work when you have got much contacts, but this "I accept the changed key" is something that many people will do without any thought - just because they don't understand the background.

**Comment by Hypolite** [comment]
> We used to be "friends", don't you find it absurd that your software gets to decide we cannot be anymore now ?

It never was about that. Of course you know you're yourself, but only through Friendica your friends have no way to be sure, and public-private keys are a way to ensure that you always are yourself. You losing your private key is functionally indistinguishable from someone forging a message on your behalf from the software perspective.

**Comment by strk** [comment]
@Hypolite initial connection is made w/out any user confirming any public key fingerprint, so the user is really always just trusting the DNS system and the supposed owner of the Friendica node hosting the target friend user.

If this kind of trust is ok on the first connection, why shoudn't it be ok later ?

I think the security aspect of Friendica is being pushed beyond reasonable boundaries, especially as it gives a false sense of security. Example: plain text mail notifications containing full body of otherwise military-grade-encrypted private messages.

I don't think Friendica is a tool that can be trusted as per privacy of shared messages, and breaking usability in the name of such false security is just bad for the project (IMHO).

@annando: the problem of users taking decisions w/out thinking about them is a non-problem. ask them to think more, give them all information about what each choice implies. instruct them. don't threat your users as sheeps...

**Comment by Hypolite** [comment]
Initial impersonation is of course always an issue, but there's nothing created public-private key pairs can do about it. However, such key pairs are required to communicate with Diaspora for example, so it's not just about Friendica's security. Even if you could recreate a private key in Friendica, Diaspora contacts wouldn't be able to communicate with you anymore because we have no control over their end.

I believe trusting users with backuping their data in case of a crash precisely is not treating them like sheeps.

**Comment by strk** [comment]
This tracker is for Friendica.
Diaspora should get its own ticket (I wont' open it because I don't use Diaspora).

@Hypolite I think not giving users a choice to avoid them taking the wrong decision is threating them like sheeps.

**Comment by annando** [comment]
Like we discussed on IRC, there would be some possible way. I guess we can detect a connection request and notify the user. We could print a notification like

> You recently had a contact request from user@domain.tld who is already your contact. Maybe someone tries to trick you - or the user had accidentally deleted your contact or had reinstalled Friendica. If you think this contact request was okay, then delete the contact and try to connect to the contact again.

**Comment by strk** [comment]
I like the suggestion, except I'd take a step further and let the user click a button to `refresh the connection` (rather than requiring manual removal and re-addition of the connection).

**Comment by strk** [comment]
I was thinking, there should be code to handle "identity migration", right ? Could this kind of recovery use that same code ?

**Comment by annando** [comment]
The identity migration works with importing the backup file that users are supposed to do when creating an account :-)

**Comment by strk** [comment]
uhm. But it still works with a different domain ?

**Comment by annando** [comment]
Yeah, because we keep the key.

**Comment by strk** [comment]
So basically only the key identifies users, not the domain.
You are not <your>@<webfinger> but a public key which is never
really shown to your contacts.

I think this is bad design. If I'm a key then:

  1. I should be able to revoke and change it, shall it be
     not secure anymore (private key leaked)

  2. I should be able to gather signatures to build trust
     (ala GPG)

  3. My contacts should clearly see my key (and for convenience
     its fingerprint) when prompted to accept my connection

  4. I shall not be recognized by domain when I ask for a new
     connection. "You are already connected" makes no sense when
     my key is different, even if my domain is the same.

The last point is what prevents me to re-connect to my previosuly
connected parties.

**Comment by strk** [comment]
This is not a `Docs`/`UX` issue, btw

**Comment by strk** [comment]
@annando  had you found any time to move steps in the direction you envisioned ?

**Comment by annando** [comment]
No. Maybe at the Hackathon somewhere at the end of the year we will find some time to discuss about this.

**Comment by MrPetovan** [comment]
@annando Do you have an estimated timeline for this feature?

**Comment by annando** [comment]
Not until now.

**Comment by MrPetovan** [comment]
This means you do have now a timeline for this feature. Maybe you were trying to say "None so far"?

**Comment by annando** [comment]
I mostly don't have timelines for any feature. I know that I really have to complete the whole item handling at first. Then I can have a look at possibly using the Diaspora transport layer to transport Friendica payload. This would fix several things. Next could be AP.

Problem with this issue here is that if really only touches the case in the Friendica <-> Friendica communication. The communication with Diaspora users will stay blocked for eternity. We cannot change their key storage method. Because of this it would be much easier to use a different hostname and/or username.

**Comment by strk** [comment]
Fix what you can, deal with what you can't.

Both Friendica and Diaspora are free software, so both can be fixed.
This ticket is about Friendica anyway.

**Comment by strk** [comment]
Someone on Diaspora pointed me to previous design about identity recover from 2016:
https://the-federation.info/specs/backup-restore/
he mentioned Diaspora devs were not welcoming to such a spec so he left:
https://wk3.org/posts/4663259#36e70de9-5017-4713-8bea-4b3b5e9dc0b6

Maybe we can do better :)

**Comment by strk** [comment]
Sorry for the noise, a quick look at that specification tells me it's not the same kind of "identity recover" I'm after. That one is for the case in which you want to change your domain, in my case I'm still in control of my domain but only lost the keypair..

**Comment by tobiasd** [comment]
We have discussed the issue on the 2018 Hackathon in Berlin. This cannot be fixed in a clean way, as well when switching to the AP protocol. Hence I'll close the issue.

**Comment by strk** [comment]
Closing this issue is a big mistake. You're missing the point, and it's a pretty big point really.
